### PR TITLE
Add vanilla mux-player

### PR DESCRIPTION
**Pull request by luwes**
Begin of the mux-player web component.

### CODE REVIEW

**Review by cjpillsbury — COMMENTED** [review]
This is looking overall super cool in my opinion. Have some questions/comments/concerns (no rush to respond).

File: packages/mux-player/package.json
Comment (luwes) [reply packages/mux-player/package.json:36]: it shouldn't conflict too much, we can turn off the styling rules of eslint if needed.
https://prettier.io/docs/en/integrating-with-linters.html

I like having eslint in though to catch my mistakes quickly

File: packages/mux-player/src/utils.js
Comment (luwes) [reply packages/mux-player/src/utils.js:1]: yes I wanted to bring this up. is this something we are really set on?
is media-chrome also going to be Typescript?

Comment (cjpillsbury) [reply packages/mux-player/src/utils.js:1]: we have a lot of users who would like types, so I think we should aim for typescript unless there's a good reason not to. There have been asks for TypeScript support in media chrome, but given that @heff would also like to minimize build dependencies, one likely path forward would be either [JSDOC-style typing definitions](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) or external type definition files.

File: packages/mux-player/src/utils.js
Comment (luwes) [reply packages/mux-player/src/utils.js:109]: @cjpillsbury how do you mean? like use a custom element with slots instead of the fragment concept?

File: packages/mux-player/src/utils.js
Comment (luwes) [reply packages/mux-player/src/utils.js:135]: this roughly translates to the React.createElement concept.
see https://reactjs.org/docs/jsx-in-depth.html

the `new Renderable` class is there for a simple `instanceof` check later but it could as well return a plain object. The important thing here is that it holds a reference to the template (component) function to re-run later on.

File: packages/mux-player/src/index.js
Comment (luwes) [reply packages/mux-player/src/index.js:151]: it returns an empty PersistentFragment
`const persistentFragment = template(props) || html``;`

File: packages/mux-player/src/index.js
Comment (luwes) [reply packages/mux-player/src/index.js:369]: this is copy / paste from CustomVideoElement.js

File: packages/mux-player/src/index.js
Comment (luwes) [reply packages/mux-player/src/index.js:374]: as well from this is copy / paste from CustomVideoElement.js

File: packages/mux-player/src/utils.js
Comment (gkatsev) [reply packages/mux-player/src/utils.js:1]: I think it may be possible to generate a ts-style type def from jsdoc? Would make bridging the divide easier in that case.

Comment (luwes) [reply packages/mux-player/src/utils.js:1]: for me TS in jsdoc is really best of both worlds. 
yes, type definitions can be generated and to create a manual .d.ts file for most public API interfaces is not that hard.

the thing that annoys me with writing TS code is:

- adds noise to the code
- takes a bit longer to write and sometimes there are TS specific issues that take a chunk of your time
- runtime type errors still happen (https://blog.logrocket.com/is-typescript-worth-it/)
- reduction of bugs is less than it's hyped up to be (https://medium.com/javascript-scene/the-typescript-tax-132ff4cb175b)

I still like it but rather have it more as a linter and doc / autocompletion tool that gets out of the way.
Also a small argument could be it makes it harder for people to contribute.

File: packages/mux-video/package.json
Comment (cjpillsbury) [inline packages/mux-video/package.json:6]: a little concerned about this causing conflicts for anyone trying to use the web component in SSR envs. Gonna move forward for now but I think we'll need to do some smoke testing before cutting a release.

Comment (cjpillsbury) [reply packages/mux-video/package.json:6]: May be fine since it's `"module"` not `"main"` fwiw

File: packages/mux-player/src/utils.js
Comment (dylanjha) [reply packages/mux-player/src/utils.js:22]: We decided on using TS for everything (and eventually circling back to port MC to TS). The reasoning is:

- Everyone wants types, writing in TS is the most reliable path to having good types
- The long term benefits we get from TS for maintainability, refactoring with confidence in the future, etc.

With the momentum TS has in the overall community of open source JS projects it feels like the right decision. That's just sort of a gut feeling.

I have definitely felt all of the "annoying points" that Wes brings up, but I still favor writing everything in TS. Even with those annoying points, I think the benefits of TS are still wroth it. Particularly when it comes to long term maintenance and eventually having to do significant refactors I find that TS gives me confidence.

Since MC started as JS I'm also not opposed to having MC use JS + JS-Doc/TS comments (also to @heff 's desire of keeping that project with minimal dependencies). My concern with that would be the possibility (probability?) of JS-Doc/TS comments becoming out of sync and causing the library to generate the wrong d.ts.

Comment (luwes) [reply packages/mux-player/src/utils.js:22]: Alright sounds good. Switching it up!

**Review by dylanjha — COMMENTED** [review]
First round of feedback, overall from the DevEx/user perspective I am very happy with this progress 🎉

File: packages/mux-player/src/index.ts
Comment (gkatsev) [reply packages/mux-player/src/index.ts:96]: I think it's worth catching here. There isn't anywhere to return to based on usage on line 69, and we probably don't want the browsers to potentially log a warning.

Comment (dylanjha) [reply packages/mux-player/src/index.ts:96]: yeah, just `catch` and log is what I was thinking

Comment (dylanjha) [reply packages/mux-player/src/index.ts:96]: but either way, it would be nice to push this `autoplay` functionality down into `<mux-video>` @cjpillsbury thoughts?

File: examples/vanilla-ts-esm/public/mux-player.html
Comment (dylanjha) [reply examples/vanilla-ts-esm/public/mux-player.html:39]: Here's a public Mux playback ID that has a captions track: `g65IqSFtWdpGR100c2W8VUHrfIVWTNRen`

File: packages/mux-player/src/template.ts
Comment (heff) [inline packages/mux-player/src/template.ts:131]: This is a really good case study for media-chrome themes. We should find some time to chat through how we can do that well and align the two so future media chrome themes could be used with mux player. Unless we know of any blockers to doing that.

File: packages/mux-player/src/video-api.ts
Comment (luwes) [reply packages/mux-player/src/video-api.ts:31]: Like we talked about since it's the base class for `MuxPlayerElement` and everything on this API interface is exposed to the public, best to have this class be an opt-in for HTMLVideoElement properties and methods. Slowly adding methods and properties when they are needed.

File: packages/mux-player/src/index.ts
Comment (luwes) [reply packages/mux-player/src/index.ts:411]: @cjpillsbury and I discussed this as well and we decided it's best to compose mux-video instead of inherit (the whole forest) from mux-video. It also makes more sense if we see mux-video as one of many media providers, better if this is loosely coupled in relation to the player.

File: packages/mux-player/src/icons.ts
Comment (luwes) [inline packages/mux-player/src/icons.ts:102]: @cjpillsbury somehow media-chrome did not change the seek-offset in the `<text>` element via the `.value` class.
maybe some race condition with light to shadow dom?

Comment (cjpillsbury) [reply packages/mux-player/src/icons.ts:102]: I'll take a look.

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [reply packages/mux-player/src/index.ts:96]: I think pushing this down to `mux-video`/`mux-video-react` or `playback-core` makes sense. I'll noodle on it.

File: packages/mux-player/src/icons.ts
Comment (cjpillsbury) [reply packages/mux-player/src/icons.ts:102]: @luwes How were you testing? the configurable seek hasn't been released yet, so just want to make sure we aren't chasing our own tail.

File: packages/mux-player/src/index.ts
Comment (dylanjha) [inline packages/mux-player/src/index.ts:123]: this is fantastic 💯

File: packages/mux-player/src/video-api.ts
Comment (dylanjha) [reply packages/mux-player/src/video-api.ts:31]: yup, agreed and we discussed this sync 👍🏼

File: packages/mux-player/src/index.ts
Comment (dylanjha) [reply packages/mux-player/src/index.ts:455]: agreed 👍🏼

File: packages/mux-player/src/index.ts
Comment (dylanjha) [reply packages/mux-player/src/index.ts:123]: Questions about the mpx-dialog error state:

1. Can we have an easy way (& document this) for developers using Mux Player to simulate the error state. As a developer adding Mux Player in my application I want to be able to visually see what the error state looks like.
2. I don't recall seeing a design for an error state, or any kind of dialog. Is that something we need to loop in some design help on?

Comment (luwes) [reply packages/mux-player/src/index.ts:123]: 1. good call, I'll see if I can add this
2. yes, true I kept it really basic. would be very helpful if we had a design for this.

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:310]: we need defaults here for `forwardSeekOffset` and `backwardSeekOffset`. I **_think_** adding them here should get picked up everywhere else.

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:161]: This feels like test code. Just making sure it's not unintentionally going to be released.

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:150]: non-blocking: Listening for fatal errors feels like it should be promoted to playback-core (at least eventually).

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:153]: We may need to be a bit smarter about this. There are non-fatal errors that may end up here (e.g. the poster image failed to load, which we encountered in Demuxed and was causing issues in mux-embed.

File: packages/mux-player/src/index.ts
Comment (cjpillsbury) [inline packages/mux-player/src/index.ts:150]: blocking: we need mux-player to redispatch an error event of some sort for these errors (stream.new relies on this)

**Review by cjpillsbury — CHANGES_REQUESTED** [review]
A couple of small callouts, only one blocking (simple version should be fine for alpha, since we can iterate on it)

File: packages/mux-player/src/index.ts
Comment (luwes) [reply packages/mux-player/src/index.ts:161]: yes, it's to make testing the error dialog easier. it can stay in the code I think, doesn't have any negative effects.

File: packages/mux-player/src/index.ts
Comment (luwes) [reply packages/mux-player/src/index.ts:153]: good call, I assumed it was only called on these https://developer.mozilla.org/en-US/docs/Web/API/MediaError but that might not be the case.

**Comment by cjpillsbury** [comment]
@luwes @dylanjha I think there are only a couple of things we need here before the merge/release:
* update media-chrome dependencies to latest release
* update the CSS in mux-player to handle the styling details for icons/buttons "from the outside" (instead of depending on the "on-hold" larger icons/buttons refactor in media-chrome)
* followup basic smoke tests (I think even if we expose minor bugs, we should merge and treat those as fast follows)
Anything missing here? Any other reasons to hold off on the merge?

**Comment by dylanjha** [comment]
that sounds like a good list @cjpillsbury ^ lets get that knocked out then we can merge this

**Comment by dylanjha** [comment]
@cjpillsbury I think we're good, yeah? I'll let you do the final review and the honors of smashing the button 🙌🏼

**Pull request merged**
